### ISSUE TITLE
Refactor magics into property of kernel object

### ISSUE THREAD

**Issue by rcsenkbeil** [ibm-et/spark-kernel#10]
### Progress of refactoring

- [x] Convert the magic interface to a single method
    - How does this handle variable arguments? The developer of the magic should be able to define what the magic takes as arguments such that he/she can avoid working with something like _Any*_; so, are annotations (for meta information) needed here? Maybe just checking for the method signature via reflection matching arguments provided?
    - How does this handle the return value? Do we still use something like _MagicOutput_ to distinguish magic return values when running through the upcoming post processor?
- [x] Refactor existing magics to use new interface
    - Update tests for existing magics
- [x] Create a Dynamics class that uses the MagicLoader to execute magics
- [x] Add post processing pipeline to check results from execution and handle return structure (whether it be Text/Plain, Application/Json, etc)
    - Needed since currently magics can return other types of output due to the split in execution path (magics vs interpreter)

### Description of current state

Currently, we have magics structured in the form of IPython where there are _line_ (%) and _cell_ (%%) magics.

```
%AddJar <some url>
```

```
%%AddJar <some url>
<another url>
<a third url>
```

This structure for magics has several downsides:

1. We have a separate code execution branch that revolves around parsing "%" and "%%"
2. We have additional complexity when designing magics as there are individual methods for line and cell magics
3. We cannot execute magics through the interpreter due to problem 1
4. Magics do not have a concrete structure (such as including command line arguments with _some_ of the magics like AddJar) 

### Goal of refactoring

The end goal of the refactoring should be to have magics exposed through the kernel API interface:

```
kernel.magics.addJar(<some url>, <some other url>, ...)
```

This involves quite a bit of refactoring to both how magics work and how code execution is processed. The _ExecuteRequestHandler_ currently splits into parsing/executing magics and interpreting normal code. This refactoring will remove the magic path of execution as they will be exposed via a bound kernel variable.

A post processor will be needed to determine how to return the output from code execution. Previously, interpreter output was always treated as _text/plain_ while magic output can be anything. The post processor needs to be able to handle different return types.

This will also change how magics are written and executed. Currently, magics have _line_ and _cell_ methods that provide raw string input. The new interface should be cleaner, allowing the developer to define what the input of the magic will be and providing a better construction for potential output.

One thought is that magics can define how the post processor reacts by returning a _MagicOutput_ in the same vein as the current implementation. This method is more unrefined as the developer is directly specifying the output (although maybe we want this?).

```
def execute(...) = {
    // ...

    MagicOutput(MIMEType.PlainText -> "some result")
}
```

Another thought is to use an annotation to mark how to wrap the return of the magic. This would allow the "wrapping" to be ignored if the magic is used in the middle of code. The downside is that it is not apparent how you could return multiple pieces of data (different MIMEs) in one magic execution.

```
@MagicOutput(MIMEType.PlainText)
def execute(...) = {

    "some result"
}
```

Can pull out meta information for return type using Scala reflection and can get argument types using TypeTags in Dynamics.

**Comment by rcsenkbeil** [comment]
As per discussion with @wellecks and @cjm5325, if we are keeping the % and %% style as an option, should we follow IPython's method of listing magics and getting more information about each?

http://ipython.org/ipython-doc/dev/interactive/reference.html

> Type %magic for more information, including a list of all available magic functions at any time and their docstrings. You can also type %magic_function_name? (see below for information on the ‘?’ system) to get information about any particular magic function you are interested in.

Furthermore, IPython contains the ability to "automagic" line magics:

> If you have ‘automagic’ enabled (as it by default), you don’t need to type in the single % explicitly for line magics; IPython will scan its internal list of magic functions and call one if it exists. With automagic on you can then just type cd mydir to go to directory ‘mydir’:

>     In [9]: cd mydir
>     /home/fperez/mydir

> Note that cell magics always require an explicit %% prefix, automagic calling only works for line magics.

> The automagic system has the lowest possible precedence in name searches, so you can freely use variables with the same names as magic commands. If a magic command is ‘shadowed’ by a variable, you will need the explicit % prefix to use it

Should functionality like this be implemented as well? Probably a good idea to create separate issues if we want to pursue these changes.

**Comment by lbustelo** [comment]
automagic is separate issue and low priority.

**Comment by wellecks** [comment]
Closing with commits 8231e7efc60858c012fa2a04528a6027fee56258 and 6e39b494927a95049d60c5b306c567708e94820a.

@cjm5325 and I worked on this: 
- implemented a ```MagicExecutor``` that extends ```Dynamic```, allowing magics to be executed with ```kernel.magics.<magic_name>```
- implemented a ```MagicParser``` that substitutes `%` syntax magic invocations with ```kernel.magics``` syntax, allowing magics to be executed with `%` and `%%`
- implemented a ```PostProcessor``` to change the cell output after a ```CellMagic``` invocation
- removed ```MagicManager``` execution path in the actor system
- divided the magic template into ```LineMagic``` and ```CellMagic``` traits
    - refactored existing magics to use the new traits
    - refactored existing magics tests
- added tests for new components and refactored existing tests

For information about how to use and develop magics with the new scheme, see [here](https://github.com/ibm-et/spark-kernel/wiki/Overview-of-Magics-for-the-Spark-Kernel) and [here](https://github.com/ibm-et/spark-kernel/wiki/Guide-to-Developing-Magics-for-the-Spark-Kernel). For a list of the currently implemented magics and their usage, see [here](https://github.com/ibm-et/spark-kernel/wiki/List-of-Current-Magics-for-the-Spark-Kernel).
